Chart tooltips improvement

alice-kroch opened the pull request:
Issue #3022.

Commit:
#3022 save chart tooltips - beginning

Commit:
#3022 tooltips connector - begin

Commit:
#3022 tooltips connector - begin, remove log

Commit:
#3022 tooltips connector - fixes by remarks

scottdraves commented on core/src/main/web/outputdisplay/bko-plot/bko-plot.js:
```diff
@@ -482,6 +483,23 @@
         };
 
+        scope.pinCloseIcon = function(d) {
+          var tipid = "tip_" + d.id;
```
looks like this code was copied.  instead, make a function that you can call from each place.

Commit:
#3022 tooltips connector - fixes by remarks

kpaharelau committed:
save chart tooltips #3022

1. Open any chart notebook
2. Create some tooltips
3. Save notebook
4. Open notebook

You can see saved tooltips

Structure

,,,json
"tips": {
                        "i0_2": {
                            "id": "i0_2",
                            "idx": 0,
                            "ele": {
                                "x": 0.49523809523809526,
                                "y": 0.5119047619047619,
                                "_x": 2,
                                "_y": 6
                            },
                            "isresp": true,
                            "cx": 316.2,
                            "cy": 120.5,
                            "op": 0,
                            "sticking": true,
                            "datax": 0.49541188738269026,
                            "datay": 0.5110997732426303,
                            "scrx": 319.9999999999995,
                            "scry": 125.0281690140854
                        },
                        "i0_4": {
                            "id": "i0_4",
                            "idx": 0,
                            "ele": {
                                "x": 0.5142857142857142,
                                "y": 0.4880952380952381,
                                "_x": 4,
                                "_y": 2
                            },
                            "isresp": true,
                            "cx": 515.5,
                            "cy": 311.5,
                            "op": 0,
                            "sticking": true,
                            "datax": 0.5149113660062565,
                            "datay": 0.48777777777777775,
                            "scrx": 523.9999999999999,
                            "scry": 312.0281690140847
                        }
                    }
,,,

kpaharelau committed:
Merge branch 'master' into alice/3022

kpaharelau committed:
Merge branch 'master' into alice/3022

kpaharelau committed:
save chart tooltips #3022

remove duplicate code - new plotTip factory has been created

kpaharelau committed:
Merge branch 'master' into alice/3022

kpaharelau committed:
save chart tooltips #3022

remove duplicate code - new plotTip factory has been created

kpaharelau committed:
save chart tooltips #3022

1) make it have 8 possible attachments, not just 4 (include the middles of the edges, not just the corners).
2) the marker should never appear when the tooltip is visible due to hover, only if the user clicks and pins.
3) target of the marker should be in the coordinates of the chart so it moves with the rest of the chart when you drag and zoom.

FIXED

kpaharelau committed:
Merge branch 'master' into alice/3022

kpaharelau committed:
save chart tooltips #3022

1) the color of the marker should be the same as the outline of the tooltip (this will make it work in themes )
2) the middle-edge attachment should be chosen more often. ideally it should choose the attachment to minimize the length of the marker.

FIXED

Also marker-style class was added to the css. In current moment marker has black color for the normal theme and white - for ambiance

scottdraves committed:
remove the markers
